File: src/main/java/uk/co/eelpieconsulting/feedlistener/ExceptionHandler.java
package uk.co.eelpieconsulting.feedlistener;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.Ordered;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ResponseStatusException;
import org.springframework.web.servlet.HandlerExceptionResolver;
import org.springframework.web.servlet.ModelAndView;
import uk.co.eelpieconsulting.common.views.ViewFactory;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

@Component
public class ExceptionHandler implements HandlerExceptionResolver, Ordered {

    private final Logger log = LogManager.getLogger(ExceptionHandler.class);

    private final ViewFactory viewFactory;

    @Autowired
    public ExceptionHandler(ViewFactory viewFactory) {
        this.viewFactory = viewFactory;
    }

    @Override
    public ModelAndView resolveException(HttpServletRequest request, HttpServletResponse response, Object handler, Exception e) {
        if (e instanceof ResponseStatusException) {
            response.setStatus(((ResponseStatusException) e).getStatus().value());
            return new ModelAndView(viewFactory.getJsonView()).addObject("data", ((ResponseStatusException) e).getReason());
        }

        log.error("Returning unexpected 500 error", e);
        return new ModelAndView(viewFactory.getJsonView()).addObject("data", "500");
    }

    @Override
    public int getOrder() {
        return Integer.MIN_VALUE;
    }

}

File: src/main/kotlin/uk/co/eelpieconsulting/feedlistener/ExceptionHandler.kt
package uk.co.eelpieconsulting.feedlistener

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.apache.logging.log4j.LogManager
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.Ordered
import org.springframework.stereotype.Component
import org.springframework.web.server.ResponseStatusException
import org.springframework.web.servlet.HandlerExceptionResolver
import org.springframework.web.servlet.ModelAndView
import uk.co.eelpieconsulting.common.views.ViewFactory

@Component
class ExceptionHandler @Autowired constructor(private val viewFactory: ViewFactory) : HandlerExceptionResolver,
    Ordered {
    private val log = LogManager.getLogger(ExceptionHandler::class.java)
    override fun resolveException(
        request: HttpServletRequest,
        response: HttpServletResponse,
        handler: Any,
        e: Exception
    ): ModelAndView {
        if (e is ResponseStatusException) {
            response.status = e.statusCode.value()
            return ModelAndView(viewFactory.jsonView).addObject("data", e.reason)
        }
        log.error("Returning unexpected 500 error", e)
        return ModelAndView(viewFactory.jsonView).addObject("data", "500")
    }

    override fun getOrder(): Int {
        return Int.MIN_VALUE
    }
}